rpk: add 'rpk cluster partitions unsafe-recover'

r-vasquez opened the pull request:
Adding `rpk cluster partitions unsafe-recover`:  This command will let the user forcefully recover data adversely affected by irrecoverable nodes.

Core's PRs: 
- https://github.com/redpanda-data/redpanda/pull/15394
- https://github.com/redpanda-data/redpanda/pull/13943

Changes introduced in this PR:
- New `rpk cluster partitions unsafe-recover` command. (Usage below)
- A new field: `Partitions Pending Recovery (<Number of partitions>)` in `rpk cluster partitions balancer-status` response.

### Usage

You must pass the dead-nodes list to execute the command, the command will print the "plan" (i.e the partitions that lost majority) and has and interactive prompt to confirm you want to perform the unsafe recovery:

```
$ rpk cluster partitions unsafe-recover --from-nodes 1
NAMESPACE  TOPIC  PARTITION  REPLICA-CORE  DEAD-NODES
kafka      bar    0          [1-1]         [1]
? Confirm recovery from these nodes? Yes
Executing recovery plan...
Successfully queued the recovery plan, you may check the status by running 'rpk cluster partitions balancer-status'

$ rpk cluster partitions balancer-status 
Status:                               ready
Seconds Since Last Tick:              26
Current Reassignment Count:           0
Partitions Pending Recovery (1):      [kafka/bar/0]
```

You may skip the confirmation prompt using the `--no-confirm` flag or just get the plan (dry run) using the `--dry` flag.

### How to reproduce:

1. Create a 3 or 5 nodes cluster using the nightly build of redpanda (you may use `rpk container`: `rpk container start -n 3 --image redpandadata/redpanda-nightly:latest`
2. Create a topic with RF=1 and produce: `rpk topic create bar -r 1` && `rpk topic produce bar...`
3. Check which node is the leader with `rpk topic describe bar -p` and stop the node (`docker stop rp-node-<id>`)
Fixes #14744

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [x] v22.3.x

## Release Notes

### Features

* rpk: add `rpk cluster partitions unsafe-recover`: This command will let the user forcefully recover data adversely affected by irrecoverable nodes.

daisukebe commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,110 @@
+			fmt.Println("Executing recovery plan...")
+			err = cl.ForceRecoverFromNode(cmd.Context(), plan, nodes)
+			out.MaybeDie(err, "unable to execute recovery plan: %v", err)
+			fmt.Println("Successfully queued the recovery plan, you may check the status by running 'rpk cluster partitions balancer-status'")
```
`move-status`?

r-vasquez replied:
I didn't see any change in the move-status command, and the Core's PR added new fields in the balancer-status response: https://github.com/redpanda-data/redpanda/pull/13943

daisukebe replied:
Ah, I see, thanks!

daisukebe replied:
It looks like there are two fields being added per #13943; `partitions_pending_force_recovery_count` and `partitions_pending_force_recovery_sample`, but the `balancer-status` command doesn't care about them, am I correct? We should update https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/cli/cluster/partitions/status.go accordingly, no? @r-vasquez

r-vasquez replied:
Yes, I added them to the API response in: [src/go/rpk/pkg/adminapi/api_cluster.go L50](https://github.com/redpanda-data/redpanda/pull/15300/files#diff-26a60a6cf5837e07ef747ecbcf45ce93a0f78ee4a9601ff3f61c79da8e4651acR50) and in the rpk response in [src/go/rpk/pkg/cli/cluster/partitions/status.go](https://github.com/redpanda-data/redpanda/pull/15300/files#diff-31c4a76f84b7c4769741e452d7d678de9c61c89204dbfa798a5518128d15eff4R103) 

It looks like (last row):

```
$ rpk cluster partitions balancer-status 
Status:                               ready
Seconds Since Last Tick:              26
Current Reassignment Count:           0
Partitions Pending Recovery (1):      [kafka/bar/0]
```

daisukebe replied:
aha, overlooked it. Great stuff!

bharathv commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+		Long: `Recover unsafely from partitions that have lost majority.
+
+This command allows you to unsafely recover all data adversely affected by the 
+loss of the nodes specified in the '--from-nodes' flag. This operation is unsafe 
```
nit: perhaps would be nice to mention that it may result in a data loss (clarifying what unsafe is).

bharathv commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+				return
+			}
+			if !noConfirm {
+				confirmed, err := out.Confirm("Confirm recovery from these nodes?")
```
nit: mention here too that it may result in a data loss?

bharathv reviewed (commented):
lgtm, added @mattschumpert for any inputs on UX.

mattschumpert commented:
@r-vasquez , @bharathv  isn't 'Partitions Pending Recovery' really 'Partitions Needing Recovery'?  IIUC this will be shown for any partitions that have lost majority. Pending sounds like something already happened.

Should we call these 'Unhealthy Partitions'?  Or 'Partitions Needing Recovery'

cc @twmb

mattschumpert commented:
@bharathv , @r-vasquez 

Also I wonder if the 'from'  in '--from-nodes' really makes sense here. Under the covers we indeed move replicas 'from' nodes but from the perspective of rpk user we're performing the action 'to' a node.  (e.g. we could use just '--nodes')

twmb commented on src/go/rpk/pkg/adminapi/api_partition.go:
```diff
@@ -155,3 +169,22 @@ func (a *AdminAPI) ToggleTopicPartitions(ctx context.Context, disabled bool, nam
+	return rr, a.sendAny(ctx, http.MethodGet, path, nil, &rr)
+}
+
+// ForceRecoverFromNode Force recovers partitions from input list of nodes.
```
nit: `ForceRecoverFromNode Force` --> lowercase the second Force

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/status.go:
```diff
@@ -95,16 +95,18 @@ investigation. A few areas to investigate:
 	fmt.Printf(format, pbs.Status, pbs.SecondsSinceLastTick, pbs.CurrentReassignmentsCount)
-
+	if pbs.PartitionsPendingForceRecovery != nil && pbs.PartitionsPendingRecoveryList != nil {
+		fmt.Printf("Partitions Pending Recovery (%v):      %v\n", *pbs.PartitionsPendingForceRecovery, pbs.PartitionsPendingRecoveryList)
```
We should create a follow up issue to make this formatting easier. The only difference between this and the out column formatter is the column formatter capitalizes the left column and it only has two spaces. I think we can make that configurable.

r-vasquez replied:
This is already possible with our table writer. We can just create a table without sections and it works. :partying_face: 

I just changed the code to use that, it's way easier to follow now. Thanks for pointing this out.

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+			if f.Kind != "text" && !dry {
+				out.Die("Format type %q is only valid for dry runs (--dry)", f.Kind)
+			}
+			if h, ok := f.Help(adminapi.MajorityLost{}); ok {
```
Doesn't this need to be []adminapi.MajorityLost? I expect a panic right now because the formatting below is based on a slice.

r-vasquez replied:
Good catch. Fixed.

twmb commented on src/go/rpk/pkg/adminapi/api_partition.go:
```diff
@@ -155,3 +169,22 @@ func (a *AdminAPI) ToggleTopicPartitions(ctx context.Context, disabled bool, nam
+
+// ForceRecoverFromNode Force recovers partitions from input list of nodes.
+func (a *AdminAPI) ForceRecoverFromNode(ctx context.Context, plan []MajorityLost, deadNodes []int) error {
+	body := map[string]interface{}{
+		"dead_nodes":                  deadNodes,
+		"partitions_to_force_recover": plan,
+	}
```
I think both Rogger and I were expecting this to basically be:
* Get a list of partitions that lost majority.
* ask for those partitions to be recovered
* optionally: specify which node to recovery from, i.e. which one do we think has the most data

Why are including the following information when asking to recover:
* term
* replicas
* dead_nodes
* _as well as_ a top level dead_nodes
What does this mean?

cc @bharathv @mattschumpert @piyushredpanda

twmb replied:
Basically I would expect us _not_ to include per-partition term,replicas,dead_nodes in the POST here, and instead just include: for each partition, recover from this potential set of nodes.

mattschumpert replied:
The base case here is explicitly _focusing_ on 1 or more nodes you _know_ are dead (not the entire cluster, though of course that could happen if the balancer moves around stuck partition replicas).  The main usage scenario is ' of 1...6, nodes 3 an 5 are dead, check them for unhealthy partitions to repair, give me a proposal listing the partitions and let me confirm the destructive operation before committing, in which I need to tell the operation which nodes to execute this on.

It's importantly a two-step operation as part of the requirements.

bharathv replied:
> optionally: specify which node to recovery from, i.e. which one do we think has the most data

this is automatically done, it picks a replica with the most latest data and hydrates the remaining from that state.

> Why are including the following information when asking to recover:

We include the following information because the state of the topic/partitions can change from the time user 'GET's the information about partitions with lost majority, digests it and the submits the request to recover them..

consider the following sequence of actions.

example 1:

> (user 1) GET  partitions_with_lost_majority 
>  (user 2) delete foo
>  (user 2) create foo
>  (user 1) POST unsafe-recover --nodes 1

Now unsafe recover command force recovers an incorrect revision of foo (which is not what the user approved)

example 2.

> (user 1)  GET  partitions_with_lost_majority 
> (user 2) manual force recover foo
> (user 1)  POST unsafe-recover --nodes 1

Now unsafe recovery will force recover `foo` again even though it was already force recovered by user 2 manually

We include all the metadata information in the payload that ties the partitions to the same instances that the user submitted the approval for.


> topic_revision (not term)

This avoids example 1 above, because there is additional validation at the time of force recovery that the revision of the topic is same (it skips if the revision has advanced).

> replicas

This avoids example 2 and similar cases. If the replica set has changed from the time the user has approved, force recovery is skipped.

> as well as a top level dead_nodes

This is the same list of nodes that are included in the rpk command, this is persisted in the controller log for the record.


> dead_nodes

This is a subset of top level dead_nodes that are actually causing lost majority for a given partition. 

example:

partition replicas = [1]
dead_nodes = [1, 2]

top level dead nodes = [1, 2]
partition dead nodes = [1]

bharathv commented on src/go/rpk/pkg/adminapi/api_partition.go:
```diff
@@ -86,6 +93,13 @@ type ClusterPartition struct {
 
+type MajorityLost struct {
+	NTP           NTP       `json:"ntp,omitempty" yaml:"ntp,omitempty"`
+	TopicRevision int       `json:"topic_revision" yaml:"topic_revision"`
```
`topic_revision` is not term, it is a monotonic counter (auto assigned by the system) that can be used differentiate between two instances of same topic.

there are a bunch of revision definitions but topic_revision is unique.

eg: 

create topic foo - topic_revision = 5
delete foo
create topic foo - topic_revision = 30

twmb replied:
I see, makes sense.

bharathv commented:
> isn't 'Partitions Pending Recovery' really 'Partitions Needing Recovery'? IIUC this will be shown for any partitions that have lost majority. Pending sounds like something already happened.
> Should we call these 'Unhealthy Partitions'? Or 'Partitions Needing Recovery'
> Also I wonder if the 'from' in '--from-nodes' really makes sense here. Under the covers we indeed move replicas 'from' nodes but from the perspective of rpk user we're performing the action 'to' a node. (e.g. we could use just '--nodes')

your suggestions sgtm, --nodes = simpler.

r-vasquez commented:
@bharathv @mattschumpert If we are moving to `--nodes` then the help text could be:

```
--nodes  Nodes that made a partition to lose majority
```
?

From the user perspective, all I see are dead nodes, and I feel that `recover from nodes` is simpler to explain.

 When you say we are "performing" an action to a node, what action is this? isn't the node dead and we want to forcefully recover data adversely affected by irrecoverable (dead) nodes?

mattschumpert commented:
@r-vasquez I'm fine if we keep the --from-nodes.

See my other suggestion, though: "Partitions Needing Recovery"

r-vasquez commented:
@mattschumpert I'll defer that to @bharathv. 

As I understand, we will only see this new field `partitions_pending_force_recovery_count` after we call `POST /v1/partitions/force_recover_from_nodes` and I think these are the partitions queued for force recovery. I think before calling the POST we won't see anything there.

mattschumpert commented:
Is that true @bharathv ? in that case it would be more of a pending state.

Feediver1 commented on src/go/rpk/pkg/adminapi/api_cluster.go:
```diff
@@ -47,6 +47,13 @@ type PartitionBalancerStatus struct {
 	// CurrentReassignmentsCount is the current number of partition
 	// reassignments in progress.
 	CurrentReassignmentsCount int `json:"current_reassignments_count,omitempty"`
+	// PartitionsPendingForceRecovery Is the number of partitions that are yet
+	// to be force recovered. It's a pointer since not all Redpanda versions
+	// include this parameter.
```
```suggestion
	// PartitionsPendingForceRecovery Specifies the number of partitions that are yet
	// to be force recovered. This is a pointer because not all Redpanda versions
	// include this parameter.
```

Feediver1 commented on src/go/rpk/pkg/adminapi/api_cluster.go:
```diff
@@ -47,6 +47,13 @@ type PartitionBalancerStatus struct {
+	// include this parameter.
+	PartitionsPendingForceRecovery *int `json:"partitions_pending_force_recovery_count"`
+	// PartitionsPendingRecoveryList Is a sample list of partitions pending
+	// force recovery (limit capped to 10)
```
```suggestion
	// force recovery (limit capped to 10).
```

Feediver1 commented on src/go/rpk/pkg/adminapi/api_partition.go:
```diff
@@ -155,3 +169,22 @@ func (a *AdminAPI) ToggleTopicPartitions(ctx context.Context, disabled bool, nam
 	return a.sendToLeader(ctx, http.MethodPost, disableURL, body, nil)
 }
+
+// MajorityLostPartitions returns a list of partitions with lost majority given an input
```
```suggestion
// MajorityLostPartitions returns a list of partitions that have lost majority when given an input
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+
+This command allows you to unsafely recover all data adversely affected by the 
+loss of the nodes specified in the '--from-nodes' flag. This operation is unsafe 
+as it allows you to bulk move all partitions that have lost majority when nodes 
```
```suggestion
because it allows you to bulk move all partitions that have lost majority when nodes 
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+This command allows you to unsafely recover all data adversely affected by the 
+loss of the nodes specified in the '--from-nodes' flag. This operation is unsafe 
+as it allows you to bulk move all partitions that have lost majority when nodes 
+are gone and irrecoverable; this may result in data loss.
```
this may result in data loss.
@Deflaimun In the docs, this should be a WARNING admonition.

Feediver1 commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+as it allows you to bulk move all partitions that have lost majority when nodes 
+are gone and irrecoverable; this may result in data loss.
+
+You may perform a dry run and verify the partitions that will be recovered by 
+using the '--dry' flag.
```
```suggestion
You can perform a dry run and verify the partitions that will be recovered by 
using the '--dry' flag.
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+			fmt.Println("Executing recovery plan...")
+			err = cl.ForceRecoverFromNode(cmd.Context(), plan, nodes)
+			out.MaybeDie(err, "unable to execute recovery plan: %v", err)
+			fmt.Println("Successfully queued the recovery plan, you may check the status by running 'rpk cluster partitions balancer-status'")
```
```suggestion
			fmt.Println("Successfully queued the recovery plan, you can check the status by running 'rpk cluster partitions balancer-status'")
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+	}
+	p.InstallFormatFlag(cmd)
+
+	cmd.Flags().IntSliceVar(&nodes, "from-nodes", nil, "Comma-separated list of nodes ID to recover the partitions from")
```
```suggestion
	cmd.Flags().IntSliceVar(&nodes, "from-nodes", nil, "Comma-separated list of node IDs from which to recover the partitions")
```

r-vasquez commented:
[Last force push](https://github.com/redpanda-data/redpanda/compare/863bd927072b68d73ea1939870309856087419da..8c1efc60ebc52f68a555072de37eb559fbd1fa62):
- Solved documentation review comments (thanks!)
- Added support for the rpk command in ducktape and replaced the admin API call to use rpk instead (8c1efc60ebc52f68a555072de37eb559fbd1fa62)

r-vasquez commented:
/dt

bharathv commented:
> Is that true @bharathv ? in that case it would be more of a pending state.

That is true, that is exactly what it means in the balancer status. Honestly I don't see much difference between needing/pending from a user perspective.

r-vasquez commented:
/ci-repeat 3
skip-units
dt-repeat=10
tests/rptest/tests/partition_force_reconfiguration_test.py::PartitionForceReconfigurationTest.test_node_wise_recovery

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/unsafe_recover.go:
```diff
@@ -0,0 +1,109 @@
+	for _, p := range plan {
+		var replicas []string
+		for _, r := range p.Replicas {
+			replicas = append(replicas, fmt.Sprintf("%v-%v", r.NodeID, r.Core))
```
I thought we had a `String()` method on a replica

r-vasquez commented:
/backport v23.3.x
